port(ed?) to plasma 6

moodyhunter opened the pull request:
KDE 6 will be released tomorrow, so it's probably the time to open this PR

similar like https://github.com/psifidotos/applet-window-title/pull/153, the ported version has been running on my machine for more than 2 months without noticeable issue.

This is a breaking change to support Qt6, which means the drop of Qt5/plasma5 support.

- raised cmake_minimum_required as per required by Qt
- Use <KSvg/Svg> instead of <Plasma/Svg>
- ported to new KPluginFatory/ButtonGroup/... API
- ported some `Q*Pointer<>` to `std::*_ptr<>`
- removed `root.plasma515` property as 6 always $>$ 5

A `ColorOverlay` is removed from AuroraeButton (short lives qtgraphicaleffects!), and `QPalette::Base` is now used as `m_activeFrameColor`, no idea of what implication will be, but this does look ok.

moodyhunter committed:
qt6 try

moodyhunter committed:
update to plasma 6 beta

moodyhunter committed:
add X-Plasma-API-Minimum-Version to metadata.json

luisbocanegra committed:
Fix "Show only for windows in current screen" not working

screenGeometry now belongs to PlasmoidItem

dr460nf1r3 commented:
I think the chances of getting this merged are quite low as @psifidotos has been absent for a while. Nevertheless, its very cool to see your efforts :) 

Question: is this confirmed to work on KDE release builds? I'm asking because it compiles successfully when using Arch's `kde-unstable` builds from 18.02. ([link](https://archlinux.org/packages/kde-unstable/x86_64/plasma-desktop/)), but not when using actual KDE master branch builds of the involved packages.

This is the error I'm getting during compilation:
~~~
==> Starting build()...
-- The C compiler identification is GNU 13.2.1
-- The CXX compiler identification is GNU 13.2.1
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/c++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD
-- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Success
-- Found Threads: TRUE
-- Found OpenGL: /usr/lib/libOpenGL.so
-- Found WrapOpenGL: TRUE
-- Found XKB: /usr/lib/libxkbcommon.so (found suitable version "1.6.0", minimum required is "0.5.0")
-- Found WrapVulkanHeaders: /usr/include
-- Performing Test HAVE_STDATOMIC
-- Performing Test HAVE_STDATOMIC - Success
-- Found WrapAtomic: TRUE
-- Found KF6CoreAddons: /usr/lib/cmake/KF6CoreAddons/KF6CoreAddonsConfig.cmake (found version "6.0.0")
-- Found KF6Declarative: /usr/lib/cmake/KF6Declarative/KF6DeclarativeConfig.cmake (found version "6.0.0")
-- Installing in the same prefix as Qt, adopting their path scheme.
-- Found KF6Package: /usr/lib/cmake/KF6Package/KF6PackageConfig.cmake (found version "6.0.0")
-- Found KF6Svg: /usr/lib/cmake/KF6Svg/KF6SvgConfig.cmake (found version "6.0.0")
-- Found KF6: success (found suitable version "6.0.0", minimum required is "5.246.0") found components: CoreAddons Declarative Package Svg
-- Found X11: /usr/include
-- Looking for XOpenDisplay in /usr/lib/libX11.so;/usr/lib/libXext.so
-- Looking for XOpenDisplay in /usr/lib/libX11.so;/usr/lib/libXext.so - found
-- Looking for gethostbyname
-- Looking for gethostbyname - found
-- Looking for connect
-- Looking for connect - found
-- Looking for remove
-- Looking for remove - found
-- Looking for shmat
-- Looking for shmat - found
-- Looking for IceConnectionNumber in ICE
-- Looking for IceConnectionNumber in ICE - found
-- KDECORATION2 VERSION MINOR : 0
CMake Deprecation Warning at libappletdecoration/CMakeLists.txt:1 (cmake_minimum_required):
Compatibility with CMake < 3.5 will be removed from a future version of
CMake.

Update the VERSION argument <min> value or use a ...<max> suffix to tell
CMake that the project does not need compatibility with older versions.


-- Setting build type to 'Debug' as none was specified.
-- Looking for __GLIBC__
-- Looking for __GLIBC__ - found
-- Performing Test _OFFT_IS_64BIT
-- Performing Test _OFFT_IS_64BIT - Success
-- Performing Test HAVE_DATE_TIME
-- Performing Test HAVE_DATE_TIME - Success
-- Found KF6WindowSystem: /usr/lib/cmake/KF6WindowSystem/KF6WindowSystemConfig.cmake (found version "6.0.0")
-- Found Gettext: /usr/bin/msgmerge (found version "0.22.4")
-- Found KF6I18n: /usr/lib/cmake/KF6I18n/KF6I18nConfig.cmake (found version "6.0.0")
-- Found KF6Service: /usr/lib/cmake/KF6Service/KF6ServiceConfig.cmake (found version "6.0.0")
-- Found KF6ConfigWidgets: /usr/lib/cmake/KF6ConfigWidgets/KF6ConfigWidgetsConfig.cmake (found version "6.0.0")
-- Found KF6KCMUtils: /usr/lib/cmake/KF6KCMUtils/KF6KCMUtilsConfig.cmake (found version "6.0.0")
-- Found KF6: success (found suitable version "6.0.0", minimum required is "5.26.0") found components: CoreAddons WindowSystem I18n Service ConfigWidgets KCMUtils
-- Configuring done (5.0s)
-- Generating done (0.0s)
-- Build files have been written to: /home/nico/.cache/paru/clone/plasma6-applets-window-buttons/src/build
[  0%] Built target appletdecorationplugin_autogen_timestamp_deps
[  4%] Automatic MOC for target appletdecorationplugin
[  4%] Built target appletdecorationplugin_autogen
[  9%] Building CXX object libappletdecoration/CMakeFiles/appletdecorationplugin.dir/appletdecorationplugin_autogen/mocs_compilation.cpp.o
In file included from /home/nico/.cache/paru/clone/plasma6-applets-window-buttons/src/build/libappletdecoration/appletdecorationplugin_autogen/EWIEGA46WW/moc_previewclient.cpp:9,
from /home/nico/.cache/paru/clone/plasma6-applets-window-buttons/src/build/libappletdecoration/appletdecorationplugin_autogen/mocs_compilation.cpp:11:
/home/nico/.cache/paru/clone/plasma6-applets-window-buttons/src/build/libappletdecoration/appletdecorationplugin_autogen/EWIEGA46WW/../../../../applet-window-buttons6/libappletdecoration/previewclient.h:107:5: warning: "KDECORATION2_VERSION_MINOR" is not defined, evaluates to 0 [-Wundef]
107 | #if KDECORATION2_VERSION_MINOR >= 13
|     ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/nico/.cache/paru/clone/plasma6-applets-window-buttons/src/build/libappletdecoration/appletdecorationplugin_autogen/EWIEGA46WW/../../../../applet-window-buttons6/libappletdecoration/previewclient.h:112:5: warning: "KDECORATION2_VERSION_MINOR" is not defined, evaluates to 0 [-Wundef]
112 | #if KDECORATION2_VERSION_MINOR >= 18
|     ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/nico/.cache/paru/clone/plasma6-applets-window-buttons/src/build/libappletdecoration/appletdecorationplugin_autogen/EWIEGA46WW/../../../../applet-window-buttons6/libappletdecoration/previewclient.h:116:5: warning: "KDECORATION2_VERSION_MINOR" is not defined, evaluates to 0 [-Wundef]
116 | #if KDECORATION2_VERSION_MINOR <= 20
|     ^~~~~~~~~~~~~~~~~~~~~~~~~~
/home/nico/.cache/paru/clone/plasma6-applets-window-buttons/src/build/libappletdecoration/appletdecorationplugin_autogen/EWIEGA46WW/../../../../applet-window-buttons6/libappletdecoration/previewclient.h:117:10: error: ‘void Decoration::Applet::PreviewClient::requestShowWindowMenu()’ marked ‘override’, but does not override
117 |     void requestShowWindowMenu() override;
|          ^~~~~~~~~~~~~~~~~~~~~
make[2]: *** [libappletdecoration/CMakeFiles/appletdecorationplugin.dir/build.make:82: libappletdecoration/CMakeFiles/appletdecorationplugin.dir/appletdecorationplugin_autogen/mocs_compilation.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:351: libappletdecoration/CMakeFiles/appletdecorationplugin.dir/all] Error 2
make: *** [Makefile:146: all] Error 2
~~~

moodyhunter commented:
Hmm this is interesting, I've never tested it on KDE master branch (should be in extra-staging rn?).
Since 6.0 will be released today, I'll fix any errors once I have the updates installed.

dr460nf1r3 commented:
Yes, it's in staging already. Thanks for taking a look once you are able to.

andrevmatos committed:
fix #ifs for plasma 6.0.0 release

Several macros expected major=5 and minor>x (e.g. kdecorations 5.13 or
5.21), checking only minor, which broke with Plasma 6.0.0 stable release
(minor=0). This fixes it in some dirty way, assuming major is either 5
or 6.

moodyhunter commented:
Thanks @luisbocanegra and @andrevmatos for https://github.com/moodyhunter/applet-window-buttons6/pull/2 https://github.com/moodyhunter/applet-window-buttons6/pull/1, the compilation failure is fixed.

hazel-bunny commented:
Hi, thanks for your port. I'm using it in Fedora 40. I have an issue though, and your issue tracker is off.

[Screencast_20240301_154243.webm](https://github.com/psifidotos/applet-window-buttons/assets/73038615/7560f518-a89a-4180-80f5-efed6edb1505)

Closing child window with buttons from applet closes parent window too. It should only close child window.

Operating System: Fedora Linux 40
KDE Plasma Version: 6.0.0
KDE Frameworks Version: 6.0.0
Qt Version: 6.6.2
Kernel Version: 6.8.0-0.rc6.49.fc40.x86_64 (64-bit)
Graphics Platform: Wayland
Processors: 2 × AMD E2-9000e RADEON R2, 4 COMPUTE CORES 2C+2G
Memory: 3.7 GiB of RAM
Graphics Processor: AMD Radeon R2 Graphics
Manufacturer: HP
Product Name: HP Laptop 14-bw0xx

moodyhunter commented:
Hi, @hazel-bunny GitHub defaults not to enable Issues for forked repos, anyway, it's been enabled now. I'll try to  investigate your issue asap.

**disclaimer: I'm not familiar with applet-window-buttons right now, as my initial intention was the Qt6 porting

moodyhunter committed:
config/qml: fix decorations combobox?

moodyhunter committed:
chore: fix a few qml warnings

moodyhunter committed:
support per-screen active window tracking

moodyhunter committed:
reformat code

moodyhunter committed:
remove windowsystem.cpp/h

it's just unused

moodyhunter committed:
code cleanup

moodyhunter committed:
cmake: lift up KF6_MIN_VERSION

moodyhunter committed:
code cleanup

moodyhunter committed:
bring back aurorae theme support

moodyhunter committed:
release 0.12.0

andrevmatos committed:
fix toggle all desktops button

moodyhunter committed:
Merge pull request #6 from andrevmatos/fix/toggle-all-desktops

fix toggle all desktops button

moodyhunter committed:
always show panel during edit mode

D3SOX commented on package/metadata.json:
```diff
@@ -0,0 +1,25 @@
+{
+    "KPlugin": {
+        "Authors": [
+            {
+                "Email": "mvourlakos@gmail.com",
+                "Name": "Michail Vourlakos"
+            }
+        ],
+        "Category": "Windows and Tasks",
+        "Description": "Window buttons for your panels",
+        "Icon": "focus-windows",
+        "Id": "org.kde.windowbuttons",
+        "License": "GPLv2",
+        "Name": "Window Buttons",
+        "ServiceTypes": [
+            "Plasma/Applet"
+        ],
```
```suggestion
{
    "KPackageStructure": "Plasma/Applet",
    "KPlugin": {
        "Authors": [
            {
                "Email": "mvourlakos@gmail.com",
                "Name": "Michail Vourlakos"
            }
        ],
        "Category": "Windows and Tasks",
        "Description": "Window buttons for your panels",
        "Icon": "focus-windows",
        "Id": "org.kde.windowbuttons",
        "License": "GPLv2",
        "Name": "Window Buttons",
```

https://develop.kde.org/docs/plasma/widget/porting_kf6/#porting-an-existing-plasmoid

> the KPlugin section may still contain the ServiceTypes key. This needs to be replaced by a KPackageStructure entry in the json's top level.

moodyhunter replied:
thanks

moodyhunter committed:
Update package/metadata.json

Co-authored-by: Nico <nico@d3sox.me>

moodyhunter committed:
code cleanup

ChrTall committed:
Removed Latte-Dock specific code from the applet, because Latte Dock is not going to be ported to Plasma 6

ChrTall committed:
Split settings into two pages: General and Behavior.
Added new Setting KWin BorderlessMaximumWindows to the Behavior Page.
This setting uses a C++ QML Extension that manipulates ~/.config/kwinrc with KConfig and uses DBus to reload the KWin Config if necessary.
KWins BorderlessMaximumWindows is set by a lot of people that use this applet.

ChrTall committed:
Added License comments.

ChrTall committed:
Migrated the project to SPDX License Format.

ander-710 committed:
Update build dependencies for Ubuntu

nicolasfella committed:
Adapt to KDecoration3

Plasma 6.3 comes with a new major version of the KDecoration libary with some breaking changes

ChrTall committed:
Update config-decoration.h.cmake

Fixed build error.

moodyhunter committed:
Merge pull request #22 from nicolasfella/plasma63

Adapt to KDecoration3/Plasma 6.3

moodyhunter committed:
Merge pull request #20 from ander-710/master

Update build dependencies for Ubuntu

ChrTall committed:
Merge branch 'moodyhunter:master' into master

moodyhunter committed:
Merge pull request #18 from ChrTall/master

Removed Latte-Dock specific code from the applet, because Latte Dock …

> I added a setting to configure the KWin option BorderlessMaximizedWindwos, which makes sense, because its only real use case is given, when using this plasmoid.
> I added a new Behavior tab to the settings page, which uses Kirigami like the KDE stock plasmoids.
> The existing General page can not be converted to Kirigami Formlayout, because the Drag and Drop of the functionality of the rearrange buttons feature does not work then.
> For now I´d suggest to keep it as it is. (unlike Kirigami it is not responsive, when resizing the window)
> The licenses were converted with kde licensedigger to the recommended reuse compliant spdx format.
